bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,210 @@
+            writer.flush();
+        } else {
+            logger.debug("Unable to send message: {}. OpenThermGatewaySocketConnector is not connected.", msg);
+        }
+    }
+
+    private void handleMessage(String message) {
+        if (message.length() > 2 && message.charAt(2) == ':') {
+            String code = message.substring(0, 2);
+            String value = message.substring(3);
+
+            logger.debug("Received command confirmation: {}: {}", code, value);
+            pendingCommands.remove(code);
+            return;
+        }
+
+        for (Entry<Long, GatewayCommand> timeAndCommand : pendingCommands.values()) {
+            if (System.currentTimeMillis() > timeAndCommand.getKey() + COMMAND_RESPONSE_TIME_MILLISECONDS) {
+                logger.debug("Resending command: {}", timeAndCommand.getValue().toFullString());
+                sendCommand(timeAndCommand.getValue());
+            } else if (System.currentTimeMillis() > timeAndCommand.getKey() + COMMAND_TIMEOUT_MILLISECONDS) {
+                pendingCommands.remove(timeAndCommand.getValue().getCode());
+            }
+        }
+
+        Message msg = Message.parse(message);
+
+        if (msg == null) {
+            logger.trace("Received message: {}, (unknown)", message);
+            return;
+        } else {
+            logger.trace("Received message: {}, {} {} {}", message, msg.getID(), msg.getCode(), msg.getMessageType().toString());
+        }
+
+        if (DataItemGroup.dataItemGroups.containsKey(msg.getID())) {
+            DataItem[] dataItems = DataItemGroup.dataItemGroups.get(msg.getID());
+
+            for (DataItem dataItem : dataItems) {
+                State state = null;
+
+                switch (dataItem.getDataType()) {
+                    case FLAGS:
+                        state = OnOffType.from(msg.getBit(dataItem.getByteType(), dataItem.getBitPos()));
+                        break;
+                    case UINT8:
+                    case UINT16:
+                        state = new DecimalType(msg.getUInt(dataItem.getByteType()));
+                        break;
+                    case INT8:
+                    case INT16:
+                        state = new DecimalType(msg.getInt(dataItem.getByteType()));
+                        break;
+                    case FLOAT:
+                        state = new DecimalType(msg.getFloat());
+                        break;
+                    case DOWTOD:
+                        break;
+                }
+
+                logger.trace("  Data: {} {} {} {}", dataItem.getID(), dataItem.getSubject(),
+                    dataItem.getDataType().toString(), state == null ? "" : state.toString());
+            }
+        }
+
+        if (msg.getMessageType() == MessageType.READACK || msg.getMessageType() == MessageType.WRITEDATA) {
+            receiveMessage(msg);
+        }
+    }
+
+    private void receiveMessage(Message message) {
+        if (message != null && callback != null) {
```

Comment:
How would `callback` be null here? Isn't `callback` a non-null field?

Reply:
Leftover from before adding @NonNullByDefault. Removed.